qi-lib/flow/core/deforest-cps.rkt

```diff
@@ -0,0 +1,323 @@
+                                               range.blanket? #'range.pre-arg #'range.post-arg
+                                               range.fine? #'range.arg
+                                               range))
+    (pattern default:fsp-default
+             #:attr name #''list->cstream
+             #:attr contract #'(-> list? any)
+             #:attr prepare #'list->cstream-prepare
+             #:attr next #'list->cstream-next
+             #:attr curry (lambda (ctx name) #'(lambda (v) v)))
+    )
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  ;; Transformers
+
+  (define-syntax-class fst
+    #:attributes (next f)
+    (pattern filter:fst-filter
+             #:attr f #'filter.f
+             #:attr next #'filter-cstream-next)
+    (pattern map:fst-map
+             #:attr f #'map.f
+             #:attr next #'map-cstream-next)
+    (pattern filter-map:fst-filter-map
+             #:attr f #'filter-map.f
+             #:attr next #'filter-map-cstream-next)
+    )
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  ;; Consumers
+
+  (define-syntax-class fsc
+    #:attributes (end)
+    (pattern foldr:fsc-foldr
+             #:attr end #'(foldr-cstream-next foldr.op foldr.init))
+    (pattern foldl:fsc-foldl
+             #:attr end #'(foldl-cstream-next foldl.op foldl.init))
+    (pattern list-ref:fsc-list-ref
+             #:attr end #'(list-ref-cstream-next list-ref.pos 'list-ref.name))
+    (pattern length:fsc-length
+             #:attr end #'(length-cstream-next))
+    (pattern empty?:fsc-empty?
+             #:attr end #'(empty?-cstream-next))
+    (pattern default:fsc-default
+             #:attr end #'(cstream-next->list))
+    )
+
+  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  ;; The pass
+
+  ;; Performs deforestation rewrite on the whole syntax tree.
+  (define-and-register-deforest-pass (deforest-pass ops ctx)
+    (syntax-parse (reverse ops)
+      [(c:fsc
+        t:fst ...
+        p:fsp)
+       ;; A static runtime contract is placed at the beginning of the
+       ;; fused sequence. And runtime checks for consumers are in
+       ;; their respective implementation procedure.
+       #`(esc
+          (#,((attribute p.curry) ctx (attribute p.name))
+           (contract p.contract
+                     (p.prepare
+                      (#,@#'c.end
+                       (inline-compose1 [t.next t.f] ...
+                                        p.next)
+                       '#,(prettify-flow-syntax ctx)
+                       '#,(build-source-location-vector
+                           (syntax-srcloc ctx))))
+                     p.name
+                     '#,(prettify-flow-syntax ctx)
+                     #f
+                     '#,(build-source-location-vector
+                         (syntax-srcloc ctx)))))])))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Runtime
+
+(begin-encourage-inline
```

Comment:
This is also a long block: perhaps an internal macro `define-inline` that
expands to `(begin-encourage-inline (racket:define-inline …))` would help?